cryptoauth.md

```diff
@@ -1,163 +1,184 @@
 # CryptoAuth
 
-**Draft 1**
+**Draft 2**
 
 Authors:
 
 - Emery Hemingway (@ehmry)
+- Lars Gierth (@lgierth)
+
+CryptoAuth is a packet-oriented encryption and authentication protocol.
+It is tolerant toward packet loss and out-of-order packets,
+and also provides authorization using various challenges.
+
+- [Introduction](#introduction)
+- Protocol changes
+- Dramatization
+- Cryptographic primitives
+- Implementations
+- Handshake
+  - Protocol
+  - Packet layout
+  - Hello, RepeatHello
+  - Key, RepeatKey
+- Data
+  - Protocol
+  - Packet layout
+  - Session state rollover
+- Authorization challenges
+  - Type 0: none
+  - Type 1: password
+  - Type 2: login/password
 
 
-#### *TODO*
-- *describe the primatives*
-- *ConnectToMe*
+## Introduction
 
+CryptoAuth at its essence defines:
 
-### Abstract
+1. A handshake protocol for exchanging session secrets.
+2. The application of well known cryptographic primitives to network data on layers 2 and 3.
 
-This document describes the CryptoAuth layer of the FC stack that provides
-confidentiality and authentication between network nodes, known as CryptoAuth.
-It contains a description of CryptoAuth headers and the session handshake protocol.
+CryptoAuth makes a few assumptions:
 
+- The underlying network transport is unreliable
+  and might lose any number of packets, or transmit packets in the wrong order.
+- The remote participant's Permanent Public Key has been acquired out-of-band.
 
-## Introduction
+CryptoAuth's mode of operation is simple:
+it takes as input individual network packets, encodes them, then outputs them.
+
+1. Read one network packet and the respective remote public key.
+2. If no matching session in `established` state:
+  - start new session (`new` state) or reuse existing session (other states),
+  - encode as Handshake Packet,
+  - write packet,
+3. With existing `established` session:
+  - encode as Data Packet,
+  - write packet,
+
+
+## Dramatization
+
+- **Alice:** TODO
+- **Bob:** TODO
+
+
+## Cryptographic primitives
+
+To simplify the protocol and minimize packet size overhead,
+CryptoAuth uses exactly one combination of key exchange, stream cipher, and HMAC.
+With no negotiation needed, a whole class of vulnerabilities disappear.
+
+- Key exchange: curve25519
+- Stream cipher: salsa20
```

Comment:
Yes it is xsalsa20, it takes 192bits (24bytes) nonce instead of 64bit and you can see 24 bytes of nonce being sent.